kale/loaddata/multi_domain.py

```diff
@@ -457,20 +456,70 @@ def is_valid_file(x: str) -> bool:
+    def __init__(
+        self, data_access: dict, domain_to_idx: dict, return_domain_label: Optional[bool] = False,
+    ):
+        self.domain_to_idx = domain_to_idx
+        self.data = []
+        self.labels = []
+        self.domain_labels = []
+        for domain_ in domain_to_idx:
+            n_samples = data_access[domain_].data.shape[0]
+            for idx in range(n_samples):
+                x, y = data_access[domain_][idx]
+                self.data.append(x)
+                self.labels.append(y)
+                self.domain_labels.append(domain_to_idx[domain_])
+
+        self.data = torch.stack(self.data)
+        self.labels = torch.tensor(self.labels)
+        self.domain_labels = torch.tensor(self.domain_labels)
+        self.return_domain_label = return_domain_label
+
+    def __getitem__(self, index: int) -> Tuple:
+        if self.return_domain_label:
+            return self.data[index], self.labels[index], self.domain_labels[index]
+        else:
+            return self.data[index], self.labels[index]
+
+    def __len__(self):
+        return len(self.labels)
+
+
+class MultiDomainAccess(DatasetAccess):
+    def __init__(self, data_access: dict, n_classes, return_domain_label: Optional[bool] = False):
+        super().__init__(n_classes)
+        self.data_access = data_access
+        self.domain_to_idx = {list(data_access.keys())[i]: i for i in range(len(data_access))}
+        self.return_domain_label = return_domain_label
+
+    def get_train(self):
+        train_access = {domain_: self.data_access[domain_].get_train() for domain_ in self.domain_to_idx}
+        return ConcatMultiDomainAccess(train_access, self.domain_to_idx, self.return_domain_label)
+
+    def get_test(self):
+        test_access = {domain_: self.data_access[domain_].get_test() for domain_ in self.domain_to_idx}
+        return ConcatMultiDomainAccess(test_access, self.domain_to_idx, self.return_domain_label)
+
+    def __len__(self):
+        return len(self.get_train()) + len(self.get_test())
+
+
 class MultiDomainAdapDataset(DomainsDatasetBase):
     """The class controlling how the multiple domains are iterated over.
 
     Args:
-        data_access (MultiDomainImageFolder): Multi-domain data access.
+        data_access (MultiDomainImageFolder, or MultiDomainAccess): Multi-domain data access.
         val_split_ratio (float, optional): Split ratio for validation set. Defaults to 0.1.
         test_split_ratio (float, optional): Split ratio for test set. Defaults to 0.2.
         random_state (int, optional): Random state for generator. Defaults to 1.
     """
 
     def __init__(
-        self, data_access: MultiDomainImageFolder, val_split_ratio=0.1, test_split_ratio=0.2, random_state: int = 1,
+        self, data_access, val_split_ratio=0.1, test_split_ratio=0.2, random_state: int = 1,
     ):
-        self.domain_labels = np.array(data_access.domain_labels)
+        # self.domain_labels = np.array(data_access.domain_labels)
```

Comment:
Still useful to keep it as a comment? If yes, explain a bit when to uncomment it.